### PR TITLE
Jschart cleanup/optimization/fixes

### PR DESCRIPTION
**Pull request by k-rister**
There are a bunch of minor improvements in this code as well as an added feature for histograms and a fix for Issue #249

### CODE REVIEW

**Comment by k-rister** [comment]
I added a fix for #259

**Comment by ndokos** [comment]
@k-rister: I'm getting errors about d3_queue being undefined. So I changed d3_queue.queue(N)  to d3.queue(N) in three places in jschart.js and that seems to work. Did something change recently?

**Comment by k-rister** [comment]
@ndokos This looks like a change in version 3.0.0 which was released in early June:

https://github.com/d3/d3-queue/commit/9f69e73a58baced320bf3a505c836db082900730

I suspect this is a casuality of npm giving the latest version which may not be backwards compatible.  Should we try and enforce a targeted version that we know works?  That appears to be supported:

https://docs.npmjs.com/cli/install

**Comment by portante** [comment]
So do we install a shared version of d3 for all v0.x versions of the charting interfaces or do we have one per version?  If the latter is true, then we should pick a targetted version.

**Comment by k-rister** [comment]
@portante This particular problem is related to d3-queue, not d3 itself.  From a quick glance (I should check this out more...) it looks like for d3 version 4 that d3-queue is now a module.  I believe it will come bundled with d3 or can still be installed independently (I think the syntax is the same either way).  In any event, from a historical perspective there were 2 different packages that were independently installed.

Assuming that is correct, I can update the code to use the new d3 v4 d3-queue module (new syntax, single package dependency) or we can update the rpm to installed an older d3-queue version that is hard coded.  It's probably a similar amount of work either way, I'm more disappointed by the fact that it's being broken in this manner which just makes me think it's likely to happen again if we don't hard code a version.

**Comment by ndokos** [comment]
AFAIK, there is a separate directory for v0.2 and v0.3 and separate copies of everything, so we could go the targeted way. But then we should target *all* packages, not just d3_queue.

For d3_queue, do we want to target the old version or make the change in jschart and target the new version? At some point, we'll probably want to do the latter, so the question is whether we want to do that  *now*.

**Comment by k-rister** [comment]
If we are going to hard code versions I'm inclined to at least get on the current levels before doing that (assuming it's relatively smooth other than some small syntax changes in the queue code).  Anybody disagree?

**Comment by ndokos** [comment]
I modified the spec file to hardwire the following versions:

npm install d3@3.5.17
npm install nvd3@1.8.2
npm install fastdom@1.0.3
npm install function-bind@1.1.0
npm install save-svg-as-png@1.0.1

and I'n longer installing d3_queue explicitly. @k-rister, can you go ahead and commit the global name change (d3_queue --> d3)?

**Comment by k-rister** [comment]
@ndokos Is it this simple:

[d3.queue.patch.txt](https://github.com/distributed-system-analysis/pbench/files/328562/d3.queue.patch.txt)

**Comment by k-rister** [comment]
@ndokos It doesn't appear to me that we will get d3.queue in D3 v3.5.17, only in D3 v4.* which isn't released yet:

D3 v3 API reference (w/o queue):

[https://github.com/d3/d3/wiki/API-Reference](url)

D3 v4 API reference (w/ queue):

[https://github.com/d3/d3/blob/master/README.md#api-reference](url)

Here is a v3 to v4 changes document (I will be reading this...):

[https://github.com/d3/d3/blob/master/CHANGES.md](url)

**Comment by ndokos** [comment]
@k-rister : re: the patch - yes, it's that simple.

Re. versions: I tried 3.5.17 d3 and left out d3-queue  in the npm installs and it seemed to work, but I did not check the docs. Let me try again more carefully to make sure it works - but we'll probably want to be as safe as possible, so I might reintroduce the d3-queue install in any case.

**Comment by k-rister** [comment]
So I perused the changes document (it is still a work in progress, but complete enough) and it looks like v4 is a major rewrite of d3 so I don't think we want to be picking that up.  The API is going to be completely different with a lot of behavioral differences.  Assuming that v3.5.17 does not have the d3.queue code in it, I don't think we want to change anything other than set the versions we want for the various packages (including both d3 and d3-queue).

**Comment by ndokos** [comment]
OK - leaving out d3-queue does *not* work: the spec file did not clean up properly on erase and left the old d3-queue lying around. I've added it back at version 3.0.1 (whcih requires the d3_queue->d3 renaming in jschart.js). I 've built a test RPM is in the COPR pbench-test repo: 
pbench-web-server-0.3-16geb27fe4.src.rpm, built from a branch in my github repo, consisting of the bits in https://github.com/k-rister/pbench/tree/jschart.cleanup-optimization-fixes, plus the d3_queue --> d3 renaming: https://github.com/ndokos/pbench/commit/eb27fe4e1dc85ee4d4c8799fc779f6eae6e76571  - it seems to work fine.

Before installing, I'd recommend ```dnf erase pbench-web-server; rm -rf /opt/pbench-web-server``` to get rid of all traces of the old one, since the old RPM did not clean up properly. Then install the test version with ```dnf install --enablerepo=copr-pbench-test pbench-web-server``` (although the name of the COPR repo may be different in your case).

**Comment by ndokos** [comment]
@k-rister: we *do* need the d3_queue--> d3 renaming with version 3.5.17 of d3 and 3.0.1 of d3_queue.

Try the current test RPM on COPR: pbench-web-server-0.3-17gbc34df8.src.rpm

**Comment by k-rister** [comment]
Ok, I can add that in easily enough.

**Comment by ndokos** [comment]
Thanks: the test RPM is pbench-web-server-0.3-18g66c6996.src.rpm on COPR pbench-test - seems to work fine, so I'll install it on the production server. I plan to merge this over the weekend in preparation for v0.39. Let me know if there is a problem and we need to revert the installation.
